src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1575,6 +1575,118 @@
     "     check ((expire is not null and updated is not null)"
     "             or (expire is null and updated is null)))"]))
 
+(defn change-reports-hash-to-summary-hash
+  "Changes the reports table primary key hash to remove the events from
+  the hash calculation. This allows events to be garbage collected separately
+  from reports"
+  []
+  (jdbc/do-commands
+   ["CREATE TABLE reports_transform ("
+    "    id bigint not null,"
+    "    hash bytea not null,"
+    "    transaction_uuid uuid,"
+    "    certname text not null,"
+    "    puppet_version text not null,"
+    "    report_format smallint not null,"
+    "    configuration_version text not null,"
+    "    start_time timestamp with time zone not null,"
+    "    end_time timestamp with time zone not null,"
+    "    receive_time timestamp with time zone not null,"
+    "    noop boolean,"
+    "    environment_id bigint,"
+    "    status_id bigint,"
+    "    metrics_json json,"
+    "    logs_json json,"
+    "    producer_timestamp timestamp with time zone not null,"
+    "    metrics jsonb,"
+    "    logs jsonb,"
+    "    resources jsonb,"
+    "    catalog_uuid uuid,"
+    "    cached_catalog_status text,"
+    "    code_id text,"
+    "    producer_id bigint,"
+    "    noop_pending boolean,"
+    "    corrective_change boolean,"
+    "    job_id text)"])
+
+  (jdbc/call-with-query-rows
+   ["select
+       id, transaction_uuid, certname, puppet_version, report_format, configuration_version,
+       start_time, end_time, receive_time, noop, environment_id, status_id, metrics_json,
+       logs_json, producer_timestamp, metrics, logs, resources, catalog_uuid, cached_catalog_status,
+       code_id, producer_id, noop_pending, corrective_change, job_id
+     from reports"]
+   (fn [rows]
+     (reduce (fn [hashes-seen row]
+               (conj hashes-seen
+                     (let [hash (-> (hash/report-identity-hash row)
+                                    (sutils/munge-hash-for-storage))]
+                       (when-not (contains? hashes-seen hash)
+                         (jdbc/insert-multi! "reports_transform"
+                                             (list (assoc row :hash hash))))
+                       hash)))
+             #{}
+             rows)))
```

Comment:
Overall I think this (and the pr) looks good, though I had wondered if we might want to handle the insert in chunks.  That said, doing so would make the code more complicated if it doesn't matter, and we'd need to be careful not to make the partition size too big relative to the row sizes (cf. potentially giant factsets when dealing with facts).

In any case, here's the kind of thing I was pondering (untested, and could be structured other ways):

```clojure
(letfn [(insert-row-chunk [hashes-seen chunk-rows]
          (let [[hashes-seen chunk-rows]
                (loop [hashes-seen hashes-seen
                       remainder chunk-rows
                       result []]
                  (if-not (seq remainder)
                    [hashes-seen result]
                    (let [hash (-> (hash/report-identity-hash row)
                                   sutils/munge-hash-for-storage)]
                      (if (hashes-seen hash)
                        (recur hashes-seen (rest remainder) result)
                        (recur (conj hashes-seen hash)
                               (rest remainder)
                               (conj result (assoc row :hash hash)))))))]
            (jdbc/insert-multi! "reports_transform" chunk-rows)
            hashes-seen))]
  (reduce insert-chunk #{} (partition 100 rows)))
```

Reply:
I'm not sure this makes much of a difference in the long run - the migration function starts a single transaction, and this code doesn't appear to be hanging on to the head. transaction: https://github.com/puppetlabs/puppetdb/blob/master/src/puppetlabs/puppetdb/scf/migrate.clj#L1726-L1739

I'd rather keep it mess complicated if possible.

Reply:
Yeah, the added complication may not be worth it, particularly since it sounds like the runtime was about the same as it was in the earlier version that didn't have the set and was partitioned.  So I think I'm fine sticking with the simpler version.